File: RRcoeff_to_varComp_function.R
rm(list=ls(all=TRUE))
#===============================================================================#
RRcoefToVarComponent <- function(times,numCoeff, CoeffMatrix, LegendrePolynomials) {
  varCovr=matrix(NA, ncol=times*numOfTraits, nrow=times*numOfTraits)
  k1 = 1; k2 = times
  z1 = 1; z2 = numCoeff
  for (i in 1:(numOfTraits)){
    k3 = 1; k4 = times
    z3 = 1; z4 = numCoeff
    for (j in 1:(numOfTraits)){
      varCovr[(k1:k2),(k3:k4)] = LegendrePolynomials %*% CoeffMatrix[(z1:z2),(z3:z4)] %*% t(LegendrePolynomials)
      k3 = k3+times;    k4 = k4+times
      z3 = z3+numCoeff; z4 = z4+numCoeff
    }
    k1 = k1+times;    k2 = k2+times
    z1 = z1+numCoeff; z2 = z2+numCoeff
  }
  return(varCovr)
}

#===============================================================================#
Phi = as.matrix(read.table("Legendre_polynomials.txt"))
#===============================================================================#

# 1. One trait
RRcoeff = as.matrix(read.table("3_RR_coefficients_for_1_traits.txt", header = F))
#===============================================================================#
numCoeff = ncol(Phi)  # Number of coefficients (i.e., order of fit)
timePoints = nrow(Phi) # Number of time points (i.e, number of times of repeated records)
#===============================================================================#
numOfTraits =nrow(RRcoeff)/numCoeff # Number of traits
#===============================================================================#
varCovr = RRcoefToVarComponent(times=timePoints,numCoeff=numCoeff,CoeffMatrix=RRcoeff, LegendrePolynomials=Phi)
varCovr
correlation = cov2cor(varCovr)
correlation


# 2. Two traits
RRcoeff = as.matrix(read.table("3_RR_coefficients_for_2_traits.txt", header = F))
#===============================================================================#
numCoeff = ncol(Phi)
timePoints = nrow(Phi)
#===============================================================================#
numOfTraits =nrow(RRcoeff)/numCoeff
#===============================================================================#
varCovr = RRcoefToVarComponent(times=timePoints,numCoeff=numCoeff,CoeffMatrix=RRcoeff, LegendrePolynomials=Phi)
varCovr
correlation = cov2cor(varCovr)
correlation

File: RRcoeff_to_varComp_function.py
import numpy as np
import pandas as pd

# Define the function
def RRcoefToVarComponent(times, numCoeff, CoeffMatrix, LegendrePolynomials, numOfTraits):
    varCovr = np.full((times * numOfTraits, times * numOfTraits), np.nan)
    k1, k2 = 0, times
    z1, z2 = 0, numCoeff
    for i in range(numOfTraits):
        k3, k4 = 0, times
        z3, z4 = 0, numCoeff
        for j in range(numOfTraits):
            varCovr[k1:k2, k3:k4] = LegendrePolynomials @ CoeffMatrix[z1:z2, z3:z4] @ LegendrePolynomials.T
            k3 += times
            k4 += times
            z3 += numCoeff
            z4 += numCoeff
        k1 += times
        k2 += times
        z1 += numCoeff
        z2 += numCoeff
    return varCovr

# Function to convert var/covar to correlations
def cov2cor(cov):
    stddev = np.sqrt(np.diag(cov))
    outer_stddev = np.outer(stddev, stddev)
    correlation = cov / outer_stddev
    correlation[cov == 0] = 0
    return correlation

# ================= Examples ==========================================#
# ------------ Single trait ---------------------
# Read Legendre_polynomials
Phi = pd.read_csv("Legendre_polynomials.txt", sep='\\s+', header=None).values

# Read RR coefficients:
RRcoeff = pd.read_csv("3_RR_coefficients_for_1_traits.txt", sep='\\s+', header=None).values

# Define parameters:
timePoints = Phi.shape[0]
numCoeff = Phi.shape[1]
numOfTraits = RRcoeff.shape[0] // numCoeff  

# Compute variance-covariance matrix
varCovr = RRcoefToVarComponent(timePoints, numCoeff, RRcoeff, Phi, numOfTraits)
print("Variance-Covariance Matrix:\n", varCovr)

# Compute correlation matrix
correlation = cov2cor(varCovr)
print("\nCorrelation Matrix:\n", correlation)


#======================================================#
# ------------ Two traits ---------------------
# Read Legendre_polynomials
Phi = pd.read_csv("Legendre_polynomials.txt", sep='\\s+', header=None).values

# Read RR coefficients:
RRcoeff = pd.read_csv("3_RR_coefficients_for_2_traits.txt", sep='\\s+', header=None).values

# Define parameters:
timePoints = Phi.shape[0]
numCoeff = Phi.shape[1]
numOfTraits = RRcoeff.shape[0] // numCoeff  

# Compute variance-covariance matrix
varCovr = RRcoefToVarComponent(timePoints, numCoeff, RRcoeff, Phi, numOfTraits)
print("Variance-Covariance Matrix:\n", varCovr)

# Compute correlation matrix
correlation = cov2cor(varCovr)
print("\nCorrelation Matrix:\n", correlation)
# =====================================================================#